[RISC-V][GC] Use 128gb for regions_range on RISCV64

t-mustafin opened the pull request:
Use 128gb for regions_range cause RISCV64 SV39 memory layout allows to use only 256gb of virtual memory: https://docs.kernel.org/riscv/vm-layout.html#risc-v-linux-kernel-sv39.

mangod9 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -45395,8 +45395,13 @@ HRESULT GCHeap::Initialize()
         }
         else
         {
+#ifdef TARGET_RISCV64
```
if this is expected for more platforms, might make sense to `#ifdef` a constant for the size in GB, instead of the full block with all the calculations to minimize duplication.

Maoni0 replied:
yep, for example, this could be defined after the block of code that defines `INITIAL_ALLOC`.

t-mustafin replied:
Thanks!

gbalykov commented on src/coreclr/gc/gc.cpp:
```diff
@@ -45395,8 +45395,13 @@ HRESULT GCHeap::Initialize()
         else
         {
+#ifdef TARGET_RISCV64
+            // For RISC-V Linux Kernel SV39 memory layout virtual memory limit is 256gb, so regins_range should be less
```
nit: typo `regins_range`

mangod9 reviewed (approved):
Nit: might even be easier to make it `REGIONS_RESERVATION_SIZE_GB` to reduce the `1024 *` duplication.

cshung commented:
@janvorli, is that 256GB limit something we can deduce from the operating system so that we don't need to hardcode it?

janvorli commented:
@cshung unfortunately there is no API to get virtual memory size limit on Linux. However, it seems it would be better to use the value returned by the `GCToOSInterface::GetVirtualMemoryLimit` derived value as a limit and modify the Unix implementation of the method. Currently it looks like this:
https://github.com/dotnet/runtime/blob/e121d77db3dbf7e1196cccea1fb250be5038419a/src/coreclr/gc/unix/gcenv.unix.cpp#L1080-L1092
@t-mustafin we should also modify the GlobalMemoryStatusEx implementation in PAL here:
https://github.com/dotnet/runtime/blob/e121d77db3dbf7e1196cccea1fb250be5038419a/src/coreclr/pal/src/misc/sysinfo.cpp#L473-L478

t-mustafin committed:
Set virtual memory limit for RISCV64 SV39

Use 128gb for regions_range cause RISCV64 SV39 memory layout allows to use only 256gb of virtual memory: https://docs.kernel.org/riscv/vm-layout.html#risc-v-linux-kernel-sv39.

t-mustafin commented:
@janvorli `GCToOSInterface::GetVirtualMemoryLimit()` returns 128TB for non-RISCV archetictures instead of 256GB used by `gc_heap::regions_range`. To avoid magic constants in code I propose to use half of `GCToOSInterface::GetVirtualMemoryLimit()` to regions mapping.

And I also added RISCV ifdef for `GlobalMemoryStatusEx`.

cshung commented on src/coreclr/gc/gc.cpp:
```diff
@@ -45747,8 +45747,8 @@ HRESULT GCHeap::Initialize()
-            // If no hard_limit is configured the reservation size is max of 256gb or 2x physical limit
-            gc_heap::regions_range = max(((size_t)256 * 1024 * 1024 * 1024), (size_t)(2 * gc_heap::total_physical_mem));
+            // If no hard_limit is configured the reservation size is max of half of GetVirtualMemoryLimit() or 2x physical limit
+            gc_heap::regions_range = max(GCToOSInterface::GetVirtualMemoryLimit()/2, (size_t)(2 * gc_heap::total_physical_mem));
```
That `max` was intended for large machines with more memory than 256G. We have got a bug report on customer running services on a machine with 750G memory unable to use all its RAM.

t-mustafin replied:
Thanks.

cshung commented on src/coreclr/gc/gc.cpp:
```diff
@@ -45747,8 +45747,8 @@ HRESULT GCHeap::Initialize()
-            // If no hard_limit is configured the reservation size is max of 256gb or 2x physical limit
-            gc_heap::regions_range = max(((size_t)256 * 1024 * 1024 * 1024), (size_t)(2 * gc_heap::total_physical_mem));
+            // If no hard_limit is configured the reservation size is max of half of GetVirtualMemoryLimit() or 2x physical limit
+            gc_heap::regions_range = max(GCToOSInterface::GetVirtualMemoryLimit()/2, (size_t)(2 * gc_heap::total_physical_mem));
```
```suggestion
            gc_heap::regions_range = max(min(GCToOSInterface::GetVirtualMemoryLimit()/2, (size_t)256 * 1024 * 1024 * 1024), (size_t)(2 * gc_heap::total_physical_mem));
```

t-mustafin replied:
Thanks.

cshung replied:
My code is actually wrong. In case the machine happen to have physical memory more than the address space size. This can be a problem.

Can we change it to `min(GCToOSInterface::GetVirtualMemoryLimit()/2, max((size_t)256 * 1024 * 1024 * 1024), (size_t)(2 * gc_heap::total_physical_mem))` instead?

Sorry for the inconvenience.

t-mustafin committed:
PR feedback

janvorli reviewed (approved):
LGTM, thank you!

gbalykov reviewed (commented):
LGTM
